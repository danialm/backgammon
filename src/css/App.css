.App {
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 80px;
}

.App-header {
  background-color: #222;
  height: 150px;
  padding: 20px;
  color: white;
}

.App-intro {
  font-size: large;
}

@keyframes App-logo-spin {
  from { transform: rotate(0deg); }
  to { transform: rotate(360deg); }
}

.turn {
  display: inline-block;
  width: 400px;
  margin-bottom: 20px;
}

.board {
  width: 264px;
  padding: 20px;
  background: white;
  display: inline-block;
}

.points {
  margin: 0;
  padding: 0;
  display: flex;
  flex-wrap: wrap;
}

.point {
  display: inline-block;
  width: 7.7%;
  height: 150px;
  position: relative;
  border-top: 1px black solid;
  border-radius: 8px;
  box-sizing: border-box;
}

.point.selected .checker:last-child{
  opacity: .3;
}

.point.possible{
  background-color: rgba(0, 152, 0, 0.16);
}

.point:nth-child(1){
  border-right: solid 1px black;
  order: 12;
}

.point:nth-child(2){
  order: 11;
}

.point:nth-child(3){
  order: 10;
}

.point:nth-child(4){
  order: 9;
}

.point:nth-child(5){
  order: 8;
}

.point:nth-child(6){
  order: 7;
  border-left: solid 1px black;
  margin-left: 7.6%;
}

.point:nth-child(7){
  border-right: solid 1px black;
  order: 6;
}

.point:nth-child(8){
  order: 5;
}

.point:nth-child(9){
  order: 4;
}

.point:nth-child(10){
  order: 3;
}

.point:nth-child(11){
  order: 2;
}

.point:nth-child(12){
  border-left: solid 1px black;
  order: 1;
}

.point:nth-child(13){
  border-left: solid 1px black;
}

.point:nth-child(24){
  border-right: solid 1px black;
}

.point:nth-child(18){
  border-right: solid 1px black;
  margin-right: 7.6%;
}

.point:nth-child(19){
  border-left: solid 1px black;
}

.point:nth-child(-n+12) {
  border-top: 0;
  border-bottom: 1px black solid;
}

.point:nth-child(-n+12) .checkers{
  transform: rotate(180deg);
  bottom: 0;
}

.checkers {
  list-style: none;
  margin: 0;
  padding: 0;
  position: absolute;
}

.checker {
  border: 1px solid black;
  border-radius: 50%;
  height: 18px;
  width: 18px;
}

.checker.b {
  background: black;
}

.dice {
  margin: 20px 0 0;
  padding: 0;
  list-style: none;
  height: 80px;
}

.die {
  display: inline-block;
  margin-right: 10px;
  width: 75px;
  height: 75px;
  background: url('../img/dice-faces.svg');
}

.die.number-1 {
  background-position: -43px 129px;
}

.die.number-2 {
  background-position: -122px 129px;
}

.die.number-3 {
  background-position: -201px 129px;
}

.die.number-4 {
  background-position: -280px 129px;
}

.die.number-5 {
  background-position: -359px 129px;
}

.die.number-6 {
  background-position: -438px 129px;
}

.hit {
  background-color: white;
  position: relative;
  left: -162px;
  top: 30px;
}

.turn, .board, .hit {
  float: left;
}

.crier {
  position: fixed;
  top: 100px;
  right: 30px;
  width: 300px;
}

.cries {
  list-style: none;
  padding: 0;
  margin: 0;
}

.cry {
  padding: 10px;
  border: 1px solid white;
}

.cry.error {
  color: red;
}

.cry.warning {
  color: orange;
}

.cry.success {
  color: green;
}

.cry.info {
  color: white;
}

.cry .cry-collapse {
  margin-right: 5px;
}
