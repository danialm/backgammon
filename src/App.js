import React, { Component } from 'react';
// import logo from './logo.svg';
import './App.css';

class Dice extends Component{
  render(){
    return(
      <ul className='dice'>
        <li className='die'>{this.props.value[0]}</li>
        <li className='die'>{this.props.value[1]}</li>
      </ul>
    )
  }
}

class Point extends Component{
  render(){
    const checkers = this.props.checkers.map((checker, i)=>{
      const klass = 'checker ' + checker;
      return(<li className={klass} key={i}></li>)
    });

    return(
      <ul className='checkers'>{checkers}</ul>
    )
  }
}

class Board extends Component{
  render(){
    const temp = Object.assign({}, this.props.points),
    points = Object.keys(temp).map((key)=>{
      const klass = 'point' + (this.props.selected === key ? ' selected' : '')
      return(
        <li
          key={key}
          className={klass}
          onClick={() => this.props.onClick(key)} >
          <Point checkers={temp[key]} />
        </li>
      );
    });

    return(
      <div className='board'>
        <ul className='points'>{points}</ul>
      </div>
    )
  }
}

class Game extends Component{
  constructor(){
    super();

    this.state = {
      whiteIsPlaying: true,
      white: {
        name: 'White'
      },
      black: {
        name: 'Black'
      },
      points: initPoints(),
      dice: {
        value: [null, null],
        rolled: false
      },
      selected: 0
    }
  }

  handelDiceRoll(){
    if(this.state.dice.rolled){ return false; }

    this.setState({
      dice: {
        value: [rand(), rand()],
        rolled: true
      }
    });
  }

  handleMove(key){
    console.log(key);
    if(key === this.state.selected){
      this.setState({ selected: 0 });
      return true;
    }

    const points = Object.assign({}, this.state.points),
          token = points[this.state.selected].pop();

    points[key].push(token);
    this.setState({
      points: points,
      whiteIsPlaying: !this.state.whiteIsPlaying,
      selected: 0,
      dice: {
        rolled: false,
        value: [null, null]
      }
    });
  }

  handleSelectPoint(key){
    if(!this.state.dice.rolled){ return false; }
    if(this.state.selected){ return this.handleMove(key); }

    if(this.state.whiteIsPlaying && this.state.points[key].indexOf('w') < 0){
      return false;
    }

    if(!this.state.whiteIsPlaying && this.state.points[key].indexOf('b') < 0){
      return false;
    }

    this.setState({selected: key});
  }

  render(){
    const turn = this.state.whiteIsPlaying ?
                 this.state.white.name :
                 this.state.black.name;

    return(
      <div>
        <div className='turn'>This is <strong>{turn}</strong>
          's turn.
          <a href='#' onClick={() => this.handelDiceRoll() }> Roll</a>
          <Dice value={this.state.dice.value} />
        </div>
        <Board points={this.state.points}
               selected={this.state.selected}
               onClick={ (key) => this.handleSelectPoint(key) } />
      </div>
    )
  }
}

function initPoints(){
  const points = {};

  for(let i = 1; i < 25; i++) {
    points[i] = [];
  }

  points[1] = ['w', 'w'];
  points[6] = ['b', 'b', 'b', 'b', 'b'];
  points[8] = ['b', 'b', 'b'];
  points[12] = ['w', 'w', 'w', 'w', 'w'];
  points[13] = ['b', 'b', 'b', 'b', 'b'];
  points[17] = ['w', 'w', 'w'];
  points[19] = ['w', 'w', 'w', 'w', 'w'];
  points[24] = ['b', 'b'];

  return points;
}

function rand(){
  return Math.floor(Math.random() * 6) + 1;
}

const App = function() {
  return(<Game />);
}

export default App;
