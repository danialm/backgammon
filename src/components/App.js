import React, { Component } from 'react';
import { connect } from 'react-redux';
import {
  BrowserRouter as Router,
  Route, Link, NavLink, Switch
} from 'react-router-dom';
import '../css/App.css';
import $ from 'jquery';
import { cryError, cryInfo, clearCries } from '../actions';
import Crier from './Crier.js';
import Config from './Config.js';
import LogLink from './LogLink.js';
import Game from './Game';
import Games from './Games';
import Login from './Login';
import Signup from './SignUp';
import ResetPassword from './ResetPassword';
import Home from './Home';
import Profile from './Profile';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      token: localStorage.getItem('token'),
      user: {
        email: '',
        password: '',
        confirmPassword: ''
      }
    }

    this.handleLoginSubmit = this.handleLoginSubmit.bind(this);
    this.handleLoginChange = this.handleLoginChange.bind(this);
    this.handleRegisterSubmit = this.handleRegisterSubmit.bind(this);
    this.handleLogOut = this.handleLogOut.bind(this);
    this.handleChangePassword = this.handleChangePassword.bind(this);
    this.fetchSelf = this.fetchSelf.bind(this);
  }

  componentWillMount() {
    if(this.state.token){
      this.fetchSelf();
    }
  }

  fetchSelf(success) {
    const t = this;
    $.ajax({
      url: Config.serverUrl + 'users/me',
      method: 'GET',
      beforeSend: function(xhr){
        xhr.setRequestHeader('Authorization', 'Bearer ' + t.state.token);
      },
      success: function(data){
        t.setState({user: data});
        if(typeof success !== 'undefined'){ success(data); }
      },
      error: function(xhr){
        t.props.dispatch(
          cryError('fetchUser', `${xhr.statusText} ${xhr.responseText}`)
        );
        localStorage.removeItem('token');
        t.setState({token: null});
      }
    });
  }

  handleLoginChange(event) {
    const user = Object.assign({}, this.state.user);
    let value = event.target.value;

    if(event.target.name === 'email'){
      let valueArray = value.split('@');
      if(valueArray[1]){
        value = [valueArray[0], valueArray[1].toLowerCase()].join('@');
      }
    }

    user[event.target.name] = value;
    this.setState({user: user});
  }

  handleLoginSubmit(event) {
    event.preventDefault();
    const t = this;

    $.ajax({
      url: Config.serverUrl + 'user_token',
      data: {
        auth: {email: t.state.user.email, password: t.state.user.password}
      },
      method: 'POST',
      success: function(data){
        localStorage.setItem('token', data.jwt);
        t.props.dispatch(clearCries());
        t.setState({token: data.jwt});
      },
      error: function(xhr){
        t.props.dispatch(
          cryError('login', `${xhr.statusText} ${xhr.responseText}`)
        );
        localStorage.removeItem('token');
      }
    });
  }

  handleRegisterSubmit(event) {
    event.preventDefault();
    const t = this;

    if(t.state.user.password.trim() === '' || t.state.user.email.trim() === ''){
      t.props.dispatch(cryError('register', 'All the fields are required!'));
      return false;
    }

    if(t.state.user.password !== t.state.user.confirmPassword){
      t.props.dispatch(cryError('register', 'Passwords do not match!'));
      return false;
    }

    $.ajax({
      url: Config.serverUrl + 'users',
      data: {
        user: {
          email: t.state.user.email,
          password: t.state.user.password
        }
      },
      method: 'POST',
      success: function(data){
        t.props.dispatch(clearCries());
        // hashHistory.push('/login');
      },
      error: function(xhr){
        t.props.dispatch(
          cryError('register', `${xhr.statusText} ${xhr.responseText}`)
        );
      }
    });
  }

  handleLogOut(event) {
    event.preventDefault();
    const user = Object.assign({}, this.state.user)

    localStorage.removeItem('token');

    user.password = ''
    user.confirmPassword = ''

    this.setState({
      token: null,
      user: user
    });
  }

  handleChangePassword(event) {
    event.preventDefault();

    const t = this;

    if(!t.state.user.email) {
      t.props.dispatch(
        cryError('changePassword', 'We need your email address!')
      );
      return false;
    }

    $.ajax({
      url: Config.serverUrl + 'password/edit',
      method: 'GET',
      data: {
        user: { email: this.state.user.email }
      },
      success: function(data){
        t.props.dispatch(cryInfo(
          'resetPassword',
          'Reset token has send to your email',
          <Link to="/reset-password">rest password</Link>
        ));
      },
      error: function(xhr){
        t.props.dispatch(
          cryError('resetPassword', `${xhr.statusText} ${xhr.responseText}`)
        );
      }
    });
  }

  render() {
    const login = <Login user={this.state.user}
                         onSubmit={this.handleLoginSubmit}
                         onChange={this.handleLoginChange}
                         changePasswordHandler={this.handleChangePassword}/>
    return(
      <Router>
        <div>
          { !this.state.token && // User is not authenticated
            <div className="container">
              <Crier />
              <Switch>
                <Route path="/login" render={() => (login) } />

                <Route path="/sign-up" render={() => (
                  <Signup user={this.state.user}
                          onSubmit={this.handleRegisterSubmit}
                          onChange={this.handleLoginChange}/>
                )}/>

                <Route path="/reset-password" render={() => (
                  <ResetPassword user={this.state.user} />
                )}/>

                <Route render={() => ( login )}/>
              </Switch>
            </div>
          }
          { this.state.token && // User is authenticated
            <div>
              <ul className="nav">
                <li><NavLink to="/" exact>Home</NavLink></li>
                <li><NavLink to="/profile">Profile</NavLink></li>
                <li><NavLink to="/games" exact>Games</NavLink></li>
                <li className="log-link">
                  <LogLink token={this.state.token}
                           user={this.state.user}
                           logoutHandler={this.handleLogOut}>
                  </LogLink>
                </li>
              </ul>
              <div className="container">
                <Crier />
                <Switch>
                  <Route exact={true} path='/' component={Home}/>

                  <Route path="/profile" render={({ match }) => (
                    <Profile user={this.state.user}
                             token={this.state.token}
                             fetchSelf={this.fetchSelf}
                             changePasswordHandler={this.handleChangePassword}
                             onChange={(user) => this.setState({user: user})}/>
                  )}/>

                  <Route path="/games" exact={true} render={({ match }) => (
                    <Games user={this.state.user}
                           token={this.state.token}/>
                  )}/>

                  <Route path="/games/:id" render={({ match }) => (
                      <Game user={this.state.user}
                            token={this.state.token}
                            id={match.params.id}/>
                  )}/>

                  <Route path="/reset-password" render={({ match }) => (
                    <ResetPassword user={this.state.user}
                                   token={this.state.token}/>
                  )}/>

                  <Route component={Home}/>
                </Switch>
              </div>
            </div>
          }
        </div>
      </Router>
    );
  }
}

App = connect()(App);
export default App;
