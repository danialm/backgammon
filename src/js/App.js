import React, { Component } from 'react';
import { IndexLink, hashHistory } from 'react-router';
import '../css/App.css';
import $ from 'jquery';
import Crier from './crier.js';
import Config from './config.js';
import NavLink from './NavLink';
import LogLink from './log-link.js';

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      token: localStorage.getItem('token'),
      user: {
        email: '',
        password: '',
        confirmPassword: ''
      },
      cries: {}
    }

    this.handleLoginSubmit = this.handleLoginSubmit.bind(this);
    this.handleLoginChange = this.handleLoginChange.bind(this);
    this.handleRegisterSubmit = this.handleRegisterSubmit.bind(this);
    this.handleLogOut = this.handleLogOut.bind(this);
    this.handleCollapse = this.handleCollapse.bind(this);
    this.fetchSelf = this.fetchSelf.bind(this);
  }

  componentWillMount() {
    if(this.state.token){
      this.fetchSelf();
    }
  }

  fetchSelf(success) {
    const t = this;

    $.ajax({
      url: Config.serverUrl + 'users/me',
      method: 'GET',
      beforeSend: function(xhr){
        xhr.setRequestHeader('Authorization', 'Bearer ' + t.state.token);
      },
      success: function(data){
        t.setState({user: data});
        if(typeof success !== 'undefined'){ success(data); }
      },
      error: function(xhr){
        const cries = Object.assign({}, t.state.cries);
        cries['fetchUser'] = {
          body: 'Faild to fetch user info',
          type: 'error'
        };
        localStorage.removeItem('token');
        t.setState({
          cries: cries,
          token: null
        });
      }
    });
  }

  handleLoginChange(event) {
    const user = Object.assign({}, this.state.user);
    let value = event.target.value;

    if(event.target.name === 'email'){
      let valueArray = value.split('@');
      if(valueArray[1]){
        value = [valueArray[0], valueArray[1].toLowerCase()].join('@');
      }
    }

    user[event.target.name] = value;
    this.setState({user: user});
  }

  handleLoginSubmit(event) {
    event.preventDefault();
    const t = this;

    $.ajax({
      url: Config.serverUrl + 'user_token',
      data: { auth: t.state.user },
      method: 'POST',
      success: function(data){
        localStorage.setItem('token', data.jwt);
        t.setState({
          token: data.jwt,
          cries: {}
        });

        hashHistory.push('/games');
      },
      error: function(xhr){
        const cries = Object.assign({}, t.state.cries);
        cries['login'] = {
          body: `${xhr.statusText} ${xhr.responseText}`,
          type: 'error'
        };
        localStorage.removeItem('token');
        t.setState({cries: cries});
      }
    });
  }

  handleRegisterSubmit(event) {
    event.preventDefault();
    const t = this,
          cries = Object.assign({}, t.state.cries);

    if(t.state.user.password.trim() === '' ||
       t.state.user.email.trim() === '') {
      cries['register'] = {body: 'All the fields are required!', type: 'error'};
      t.setState({cries: cries});
      return false;
    }

    if(t.state.user.password !== t.state.user.confirmPassword){
      cries['register'] = {body: 'Passwords do not match!', type: 'error'};
      t.setState({cries: cries});
      return false;
    }

    $.ajax({
      url: Config.serverUrl + 'users',
      data: {
        user: {
          email: t.state.user.email,
          password: t.state.user.password
        }
      },
      method: 'POST',
      success: function(data){
        t.setState({cries: {}});

        hashHistory.push('/login');
      },
      error: function(xhr){
        cries['register'] = {
          body: `${xhr.statusText} ${xhr.responseText}`,
          type: 'error'
        };
        t.setState({cries: cries});
      }
    });
  }

  handleLogOut(event) {
    event.preventDefault();
    const user = Object.assign({}, this.state.user),
          cries = Object.assign({}, this.state.cries);

    localStorage.removeItem('token');

    user.password = ''
    user.confirmPassword = ''
    cries['logout'] = {body: 'success', type: 'successful'};

    this.setState({
      token: null,
      user: user,
      cries: cries
    });

    hashHistory.push('/login');
  }

  handleCollapse(event) {
    event.preventDefault();
    const cries = Object.assign({}, this.state.cries),
          key = $(event.target).closest('.cry').data('key');

    delete cries[key];

    this.setState({cries: cries});
  }

  render() {
    let props, clonedChildren;

    if(this.props.children) {
      if(this.props.children.type.name === 'Login'){
        props = {
          user: this.state.user,
          onSubmit: this.handleLoginSubmit,
          onChange: this.handleLoginChange
        }
      }else if(this.props.children.type.name === 'Signup') {
        props = {
          user: this.state.user,
          onSubmit: this.handleRegisterSubmit,
          onChange: this.handleLoginChange
        }
      }else if(['Games', 'Game'].indexOf(this.props.children.type.name) > -1 ) {
        props = {
          user: this.state.user,
          token: this.state.token
        }
      }else if(this.props.children.type.name === 'Profile') {
        props = {
          token: this.state.token,
          user: this.state.user,
          fetchSelf: this.fetchSelf,
          onChange: (user) => this.setState({user: user})
        }
      }else{
        props = {
          user: this.state.user
        }
      }

      clonedChildren = React.cloneElement(this.props.children, props);
    }else{
      clonedChildren = null;
    }

    return(
      <div>
        <ul className="nav">
          <li><IndexLink to="/" activeClassName="active">Home</IndexLink></li>
          <li><NavLink to="/profile">Profile</NavLink></li>
          <li><NavLink to="/games">Games</NavLink></li>
          <li className="log-link">
            <LogLink token={this.state.token}
                     user={this.state.user}
                     logoutHandler={this.handleLogOut}>
            </LogLink>
          </li>
        </ul>
        <Crier cries={this.state.cries} collapseHandler={this.handleCollapse} />
        { clonedChildren }
      </div>
    );
  }
}

export default App;
