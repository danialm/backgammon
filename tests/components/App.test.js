import React from 'react';
import ReactDOM from 'react-dom';
import App from '../../src/components/App';
import reducer from '../../src/reducers';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import { Route, Link, MemoryRouter, renderTestSequence } from 'react-router-dom';
import renderer from 'react-test-renderer';

const store = createStore(reducer);
const div = document.createElement('div');
class LocalStorageMock {
  constructor() {
    this.store = {};
  }

  removeItem(key) {
    this.store[key] = null;
  }

  getItem(key) {
    return this.store[key];
  }

  setItem(key, value) {
    this.store[key] = value.toString();
  }
};

global.localStorage = new LocalStorageMock;

describe('without token set', () => {
  test('renders Login', () => {
    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter initialEntries={[ '/sign-up' ]}>
          <App />
        </MemoryRouter>
      </Provider>
    );

    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
