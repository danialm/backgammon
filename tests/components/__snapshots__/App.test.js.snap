exports[`without token set renders Login 1`] = `
<div>
  <div
    className="container">
    <div
      className="crier">
      <ul
        className="cries" />
    </div>
    <div
      className="login-page">
      <h2>
        Login
      </h2>
      <form
        onSubmit={[Function]}>
        <p>
          <label>
            Email:
            <input
              name="email"
              onChange={[Function]}
              type="email"
              value="" />
          </label>
        </p>
        <p>
          <label>
            Password:
            <input
              name="password"
              onChange={[Function]}
              type="password"
              value="" />
          </label>
        </p>
        <input
          type="submit"
          value="Submit" />
      </form>
      <p>
        <span>
          Forget your password? 
        </span>
        <a
          href="#"
          onClick={[Function]}>
          Reset it!
        </a>
      </p>
      <p>
        Do not have an account? 
        <a
          href="/sign-up"
          onClick={[Function]}>
          Create One!
        </a>
      </p>
    </div>
  </div>
</div>
`;
